src/Functions/toStartOfInterval.cpp

```diff
@@ -243,15 +357,64 @@ class FunctionToStartOfInterval : public IFunction
         size_t size = time_data.size();
 
         auto result_col = result_type->createColumn();
-        auto *col_to = assert_cast<ToColumnType *>(result_col.get());
+        auto * col_to = assert_cast<ToColumnType *>(result_col.get());
         auto & result_data = col_to->getData();
         result_data.resize(size);
 
         Int64 scale_multiplier = DecimalUtils::scaleMultiplier<DateTime64>(scale);
+        Int64 scale_on_interval = decideScaleOnPrecision<unit>(scale);
+        Int64 scale_diff = scale_on_interval > scale_multiplier ? scale_on_interval / scale_multiplier : scale_multiplier / scale_on_interval;
 
-        for (size_t i = 0; i != size; ++i)
-            result_data[i] = static_cast<ToFieldType>(ToStartOfInterval<unit>::execute(time_data[i], num_units, time_zone, scale_multiplier));
+        if (origin_column.column == nullptr)
+        {
+            for (size_t i = 0; i != size; ++i)
+            {
+                result_data[i] = 0;
+                if (scale_on_interval < scale_multiplier)
+                    result_data[i] += static_cast<ToFieldType>(ToStartOfInterval<unit>::execute(time_data[i], num_units, time_zone, scale_multiplier)) * scale_diff;
+                else
+                    result_data[i] = static_cast<ToFieldType>(ToStartOfInterval<unit>::execute(time_data[i], num_units, time_zone, scale_multiplier));
+            }
+        }
+        else
+        {
+            UInt64 origin = origin_column.column->get64(0);
 
+            for (size_t i = 0; i != size; ++i)
+            {
+                auto t = time_data[i];
+                if (origin > static_cast<size_t>(t))
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "The origin must be before the end date/datetime");
+
+                t -= origin;
+                auto res = static_cast<ToFieldType>(ToStartOfInterval<unit>::execute(t, num_units, time_zone, scale_multiplier));
+
+                static constexpr size_t SECONDS_PER_DAY = 86400;
+
+                result_data[i] = 0;
+                if (unit == IntervalKind::Week || unit == IntervalKind::Month || unit == IntervalKind::Quarter || unit == IntervalKind::Year)
```

Comment:
Please add comments to l. 395-416. It is currently not clear which cases are covered, and what needs to be done in each case.